### ISSUE TITLE
APF improperly estimates cost for LIST aggregated calls

### ISSUE THREAD

**Issue by mborsz** [kubernetes/kubernetes#109106]
### What happened?

I started API call like:

```
2022-03-17 08:01:17.003 PDT
""HTTP" verb="LIST" URI="/apis/external.metrics.k8s.io/v1beta1/namespaces/ns-1/custom.googleapis.com%7Ctarget?labelSelector=resource.labels.namespace_id+in+%28default%2Cdummy%29%2Cresource.labels.pod_id%2Cresource.type%3Dgke_container"
```

Based on kube-apiserver logs I think that this call has taken `maximumSeats` for entire duration of the call (which was 10s due to reentrant call made by service handling this call and using the same priority-level).

In my opinion we shouldn't use `maximumSeats` in case of delegated calls as they do not consume much kube-apiserver resources + (practically speaking) can block for quite a long time for external reasons blocking significant part of priority level.

What happens:
* Such calls are not longrunning according to definition in https://github.com/kubernetes/kubernetes/blob/4348c8ecaf87d91503718a42a930c397c3c82569/cmd/kube-apiserver/app/server.go#L387-L390 so they are being a subject of throttling in priority and fairness
* we hit ObjectCountNotFoundErr as we don't know how many objects of type like "external.metrics.k8s.io/v1beta1" we have (which is expected), https://github.com/kubernetes/kubernetes/blob/4348c8ecaf87d91503718a42a930c397c3c82569/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/list_work_estimator.go#L70-L70 and return `maximumSeats`

### What did you expect to happen?

* InitialSeats for LIST calls to external API services is lower than `maximumSeats` (aka 10). Something like 1 is more expected as kube-apiserver is only redirecting traffic which involves limited CPU and memory, compared to other LIST calls.
* LIST calls to external API services are not handled by "error handling path", rather there is some explicit check for them

### How can we reproduce it (as minimally and precisely as possible)?

From https://github.com/kubernetes/kubernetes/issues/108524 which describes some observed problem caused by this:

* `kubectl apply -f https://raw.githubusercontent.com/GoogleCloudPlatform/k8s-stackdriver/master/custom-metrics-stackdriver-adapter/deploy/production/adapter_new_resource_model.yaml`
* `kubectl get --raw "/apis/external.metrics.k8s.io/v1beta1/namespaces/triton-batch/s0-prometheus-gpu_utilization?labelSelector=scaledobject.keda.sh%2Fname%3Dtriton"`

The second call will take 10 seats.

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by mborsz** [comment]
I think we have at least two options on the table:
* Detect in listWorkEstimator that the call we handle is for external apiservice and return smaller value
* Similarly to https://github.com/kubernetes/kubernetes/pull/105511, once we pass the request to external service, unblock queue

I prefer 1 as this avoid case when we block `maximumSeats` seats before the request starts, but as discussed with @wojtek-t it may be tricky to implement this given the current kube-apiserver/kube-aggregator architecture (the decision where to send a request happens after we estimate cost).

**Comment by liggitt** [comment]
If there's not a clear low-risk solution immediately available, is there a 1.23 change that should be reverted to resolve the regression and regain status quo in 1.23?

**Comment by wojtek-t** [comment]
> If there's not a clear low-risk solution immediately available, is there a 1.23 change that should be reverted to resolve the regression and regain status quo in 1.23?

I don't think there is a single PR that would be easily to revert.
As for low-risk mitigation - we may consider switching off this behavior and fallback to previous behavior and just simply set the width of the request always to 1.
Basically adding something like:
```
var widthEstimationEnabled = false

if !widthEstimationEnabled {
  return WorkEstimate{InitialSeats: minimumSeats}
}
```
here:
https://github.com/kubernetes/kubernetes/blob/release-1.23/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/width.go#L106

[It would require adjusting different unit tests but should be relatively mechanical]


This is effectively switching off the LIST support in P&F so if someone started relying on that we will break them.
So it's not a "no-risk" - but something we may consider it.


Regarding low-risk solution - I would like to look into it tomorrow with @mborsz so let's wait a day or two more so that we have a solution and we can then assess the risk of it (I think the solution should be low-risk, but until we have I won't be 100% sure).

**Comment by liggitt** [comment]
> The trick is how to implement delegatedCall at this point, as it depends on kube-aggregator state.

it seems like it will require breaking some of the desired API server layering to get that information up at the APF filter layer

**Comment by liggitt** [comment]
(cc @deads2k for API server layering)

**Comment by liggitt** [comment]
> This is effectively switching off the LIST support in P&F so if someone started relying on that we will break them.

unbreaking 1.22 → 1.23 upgraded users is higher priority, if we have to choose

**Comment by liggitt** [comment]
That sounds like a potentially big change to pick back to 1.23.

What would the alternative of reverting the 1.23 change to 1.22 behavior look like?

**Comment by wojtek-t** [comment]
> That sounds like a potentially big change to pick back to 1.23 (and to make at the last minute for 1.24)

I think there are other options that should mitigate them. One of them is to keep estimation as is, just "finish the request from P&F perspective" as soon as it's actually delegated to aggregated apiserver. We did something similar for the purpose of watch initialization in: https://github.com/kubernetes/kubernetes/pull/105511/files and if we ignore the test there, it's actually pretty small and in my opinion very low-risk change. I just need a couple hours to explore it deeper.
And while I agree it's still imperfect, it can serve as a very good mitigation imho.

The alternative reverting-like behavior would be what I described above in: https://github.com/kubernetes/kubernetes/issues/109106#issuecomment-1083066415

And while I agree that fixing problem is more important than not breaking customers who rely on a new behavior, that I would still prefer to achieve both to not dissatisfy the second group. So I think we need 1-2 days more to answer if any of the above can be done in small enough low-risk enough way before reverting (which given it was launched 4 months ago imho 2-3 days difference doesn't change much).

**Comment by mborsz** [comment]
I implemented a solution described in https://github.com/kubernetes/kubernetes/issues/109106#issuecomment-1083255960: https://github.com/kubernetes/kubernetes/pull/109187

The fix is quite small (in terms of lines) and more with relatively low complexity (as they are mostly mechanical).

Anyway, we discussed this with @wojtek-t and it sounds like for backport purposes partial revert of the listWorkEstimator will be a better solution (change the returned value for "err == ObjectCountNotFoundErr" case from maximumSeats to minumumSeats).

We can use the https://github.com/kubernetes/kubernetes/pull/109187 as solution in 1.24/1.25+.

**Comment by wojtek-t** [comment]
Yeah - as @mborsz wrote above, I looked into the solution I described above and this won't be cherrypickable.

I think the solution that @mborsz is proposing above is a reasonable path forward, but I actually came up with another idea (that Maciek mentioned above), which is basically to change the error handling there.
This mitigates the issue of delegated requests, without reverting the whole support for list. There will be corner cases where it will be underestimating cost of requests, but overall it's strictly better than reverting the whole support I proposed above.

In terms of risk - it's definitely a very low-risk and I will prepare a PR for that (and cherrypick to 1.23 once it merges) in the next couple hours.

**Comment by mborsz** [comment]
Also I was thinking about other potential solutions for future k8s versions and I think that we may actually want do small refactoring in master: instead of teaching listWorkEstimator about internals of all future request handlers chain, create a similar chain of EstimateWork calls that matches ServeHTTP chain:

* We have `DefaultBuildHandlerChain -> aggregator server.ServeHTTP() -> kube-apiserver.ServeHTTP() -> apiExtensions server.ServeHTTP() -> notFound.ServeHTTP()`
* We can create `DefaultBuildHandlerChain -> aggregator server.EstimateWork() -> kube-apiserver.EstimateWork() -> apiExtensions server.EstimateWork() -> notFound.EstimateWork()`

The EstimateWork functions at each level may work like:
```
func (a *apiserver) EstimateWork(req *http.Request) (WorkEstimate, error) {
  if a.IsHandledHere(req) {
     // calculate cost based on the internals I have, e.g.
     // * aggregator can return here 1
     // * "kube-apiserver with etcd" returns number based on StorageObjectCountTracker
  }
  return a.delegatedServer.EstimateWork(req)
}
```

This way we maintain apiservers layers separation, avoid passing internal knowledge (like number of objects in etcd or aggregator's urls) using Config to listWorkEstimate and in general preventing listWorkEstimate complexity growth.

**Comment by liggitt** [comment]
mitigation merged to 1.24, ready to pick to 1.23
/milestone v1.23

**Comment by wojtek-t** [comment]
The fix was cherrypicked to 1.23 in https://github.com/kubernetes/kubernetes/pull/109216

So we can close it now - it's fixed.
